pkg/tests/tasks/security/authorization/tcp_test.go

```diff
@@ -16,156 +16,194 @@ package authorization
-		time.Sleep(time.Duration(10) * time.Second)
-
-		ports := []string{"9000", "9001"}
-		for _, port := range ports {
-			cmd := fmt.Sprintf(`sh -c 'echo "port %s" | nc tcp-echo %s' | grep "hello" && echo 'connection succeeded' || echo 'connection rejected'`, port, port)
-			msg, err := util.PodExec("foo", sleepPod, "sleep", cmd, true)
-			util.Inspect(err, "Failed to get response", "", t)
-			if !strings.Contains(msg, "connection rejected") {
-				log.Log.Errorf("Verify invalid rule Unexpected response: %s", msg)
-				t.Errorf("Verify invalid rule Unexpected response: %s", msg)
-			} else {
-				log.Log.Infof("Success. Get expected response: %s", msg)
-			}
-		}
-	})
+func assertPortTcpEchoDenied(t test.TestHelper, ns string, port string) {
+	oc.Exec(t,
+		pod.MatchingSelector("app=sleep", ns),
+		"sleep",
+		fmt.Sprintf(`sh -c 'echo "port %s" | nc tcp-echo %s' | grep "hello" && echo 'connection succeeded' || echo 'connection rejected'`,
+			port, port),
+		assert.OutputContains(
+			"connection rejected",
+			fmt.Sprintf("Got expected connection rejected on port %s", port),
+			fmt.Sprintf("Expected connection rejected, but got return message hello on port %s", port)))
+}
 
-	t.Run("Security_authorization_rbac_deny_GET_tcp", func(t *testing.T) {
-		defer util.RecoverPanic(t)
-
-		log.Log.Info("Apply a DENY policy")
-		util.KubeApplyContents("foo", TCPDenyGETPolicy)
-		time.Sleep(time.Duration(10) * time.Second)
-
-		ports := []string{"9000", "9001"}
-		for _, port := range ports {
-			cmd := fmt.Sprintf(`sh -c 'echo "port %s" | nc tcp-echo %s' | grep "hello" && echo 'connection succeeded' || echo 'connection rejected'`, port, port)
-			msg, err := util.PodExec("foo", sleepPod, "sleep", cmd, true)
-			util.Inspect(err, "Failed to get response", "", t)
-
-			if port == "9000" {
-				if !strings.Contains(msg, "connection rejected") {
-					log.Log.Errorf("Verify DENY rule Unexpected response: %s", msg)
-					t.Errorf("Verify DENY rule Unexpected response: %s", msg)
-				} else {
-					log.Log.Infof("Success. Get expected response: %s", msg)
-				}
-			}
-			if port == "9001" {
-				if !strings.Contains(msg, "connection succeeded") {
-					log.Log.Errorf("Verify DENY rule Unexpected response: %s", msg)
-					t.Errorf("Verify DENY rule Unexpected response: %s", msg)
-				} else {
-					log.Log.Infof("Success. Get expected response: %s", msg)
-				}
-			}
-		}
-	})
+func assertPortTcpEchoIPAccepted(t test.TestHelper, ns string, port string) {
+	podIP := oc.GetPodIP(t,
+		pod.MatchingSelector("app=tcp-echo", ns))
+
+	oc.Exec(t,
+		pod.MatchingSelector("app=sleep", ns),
+		"sleep",
+		fmt.Sprintf(`sh -c 'echo "port %s" | nc %s %s' | grep "hello" && echo 'connection succeeded' || echo 'connection rejected'`,
+			port, podIP, port),
+		assert.OutputContains(
+			"connection succeeded",
+			fmt.Sprintf("Got expected hello message on port %s", port),
+			fmt.Sprintf("Expected return message hello, but failed on port %s", port)))
+}
```

Comment:
If, instead of getting the podIP in this function, you instead add a parameter `host string` and pass the podIP in this parameter, you can get rid of the `assertPortTcpEchoAccepted` function. Wherever you call that function, you could simply just pass `"tcp-echo"` as the `host`. 

The same goes for `assertPortTcpEchoIPDenied` and `assertPortTcpEchoDenied`. They could be a single function.

Reply:
Then, you could reduce the duplication even further by using the same pattern as I used here: 

https://github.com/maistra/maistra-test-tool/blob/e7d22a65c2d721e66f29ea1fa36e655998f65c36/pkg/tests/tasks/security/authentication/mtls_migration_test.go#L97-L117